LSPS2/README.md

```diff
@@ -0,0 +1,1001 @@
+the checks below.
+
+The LSP:
+
+* MUST check that there is currently no pending LSPS2-bought
+  channel currently being opened (i.e. has not completed up to
+  `channel_ready` from both parties).
+* MUST check that the given `opening_fee_params.promise` validates
+  the given `opening_fee_params`, and that
+  `opening_fee_params.valid_until` is a future datetime.
+* MUST check that all the PPPs listed in `ppps` are still in the
+  PPP Map.
+* MUST sum up the amounts of all PPPs listed and set `payment_size`
+  to that sum.
+* MUST compute the `opening_fee` based on the `payment_size` and
+  the given `opening_fee_params`, and MUST check that the
+  `opening_fee` is strictly less than
+  `payment_size - number_of_ppps * minimum_htlc_msat`.
+
+If all above checks pass, the call succeeds.
+Before responding, the LSP, still in the same atomic section:
+
+* Removes the PPPs listed in `ppps` from the PPP Map and puts
+  them into a PPP Opening Set.
+  * The existence of the PPP Opening Set is what is checked to
+    determine if a pending LSPS2-bought channel is being opened.
+
+Once the above operations complete, the call succeeds and the
+atomic section ends, and the LSP responds with an empty object
+`{}` and proceeds to the next step.
+
+If the checks fail, the atomic section ends and the LSP responds
+with an error code:
+
+* `operation_pending` (2) - An LSPS2-bought channel is not yet
+  completed opening and sending the PPPs being used to pay for
+  it.
+  * The client SHOULD retry later once the channel has opened and
+    the existing payments resolved.
+* `invalid_opening_fee_params` (3) - The `opening_fee_params.promise`
+  failed validation, or the `opening_fee_params.valid_until` is
+  already a past datetime.
+* `missing_ppp` (4) - One or more of the `ppps` is not in the PPP
+  Map.
+  The `data` object has a `ppps` field that contains the PPP
+  string identifiers that *are* still in the PPP Map.
+* `insufficient_funds` (5) - The total `payment_size` is not
+  enough to pay for the `opening_fee`.
+
+> **Non-normative** The intent is that once asynchronous receives
+> are possible, asyncpay-type PPPs will not be accepted in the
+> `lsps2.openwithppps` call.
+> Instead, those PPPs will be passed to a different
+> `lsps2.acceptasyncppps` call, which will cause the LSP to
+> respond to the senders of those asynchronous receives.
+> The senders would then deliver actual HTLCs/PTLCs that the LSP
+> will convert again to PPPs.
+>
+> This will help with the multiple-small-separate-payments
+> case; asyncpay-type PPPs are expected to be holdable by the
+> LSP longer, and once enough asyncpay-type PPPs are available,
+> the client can then request the senders to try sending the
+> actual payments in the hope that the LSP can gather enough
+> to pay for a new channel.
+
+### 7. Opening Channel And Forwarding Selected Pending Payment Parts
+
+The LSP opens a channel to the client.
+
+TODO: copy from existing LSPS2.
+
```

Comment:
TODOs:

The client MUST NOT use the `alias` SCID sent in the 0-conf channel open.

The LSP SHOULD fail the incoming HTLC with `unknown_next_peer` if the `alias` SCID is used as next hop.

Intent is that the client should just use this API entirely.

Reply:
Done.